hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncTable.java

```diff
@@ -233,6 +234,28 @@
    */
   CheckAndMutateBuilder checkAndMutate(byte[] row, byte[] family);
 
+  /**
+   * Atomically checks if a row matches the specified filter. If it does, it adds the
+   * Put/Delete/RowMutations.
+   * <p>
+   * Use the returned {@link CheckAndMutateBuilder} to construct your request and then execute it.
+   * This is a fluent style API, the code is like:
+   *
+   * <pre>
+   * <code>
+   * table.checkAndMutate(row).ifMatches(filter).thenPut(put)
+   *     .thenAccept(succ -> {
+   *       if (succ) {
+   *         System.out.println("Check and put succeeded");
+   *       } else {
+   *         System.out.println("Check and put failed");
+   *       }
+   *     });
+   * </code>
+   * </pre>
+   */
+  CheckAndMutateBuilder checkAndMutate(byte[] row);
```

Comment:
The API here is a bit strange. We introduce the above family & row method is because that family and row both can not be null, but here since we allow matching multiple qualifiers acrossing different families, we can accept null families as we can pass it through the filter implementation? So here, either we deprecated the old checkAndMutate(family, row) method and introduce a new family method, or we introduce a new checkAndMutate(row, filter)?

Reply:
@Apache9 Thank you for reviewing this!

> but here since we allow matching multiple qualifiers acrossing different families, we can accept null families as we can pass it through the filter implementation?

Yes. With this new feature, we can specify multiple conditions across different families by setting a filter. So the family can be null.

My intention was if we want to specify a filter by ifMatches(filter), we can use checkAndMutate(row). And if we want to specify a qualifier and a condition by qualifier()/ifNotExists()/ifEquals()/ifMatches(op, value), we can use checkAndMutate(row, family) in the existing way.

And, if we call ifMatches(filter) after calling checkAndMutate(row, family), it will throw UnsupportedOperationException with the message "Please use checkAndMutate(row)":
```
table.checkAndMutate(row, family).ifMatches(filter).thenPut(put); // UnsupportedOperationException: Please use checkAndMutate(row)
```

And vice versa:
```
table.checkAndMutate(row).qualifier(qualifier).ifMatches(op, value).thenPut(put); // UnsupportedOperationException: Please use checkAndMutate(row, family)
```

Please see the following test for this:
https://github.com/brfrn169/hbase/blob/HBASE-23146/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestCheckAndMutate.java#L380-L416

I think, you are right, it might be a little bit confusing but I'm still feeling it's okay. What do you think?

Reply:
Ping @Apache9

Reply:
@Apache9 

One of the reason why I decided the design of the new API here is that I wanted to keep the compatibility of the existing API. This patch also involves coprocessor changes, and I didn't want to change the existing methods and I just wanted to add new methods.

If you still think the API is strange, I can change it.

Thanks.

Reply:
@Apache9 Could you please reply this when you get a chance?

Reply:
Actually, this feature is needed for our use case. If we can, I want to include it in hbase-2.3.

Reply:
Sorry, we are still fighting with 2019-nCov so... Hope Japan will be safe...

I still prefer that we introduce a new method which accepts a row and filter as parameters to construct a Builder, this is more natural. Enforcing usage by method signature is better than javadoc.

Reply:
@Apache9 Thank you very much for the reply. 

```
we are still fighting with 2019-nCov so... Hope Japan will be safe...
```
Oh okay.. I hope you are well... I'm okay now, but 2019-nCov is spreading little by little in Japan actually.

```
I still prefer that we introduce a new method which accepts a row and filter as parameters to construct a Builder, this is more natural. Enforcing usage by method signature is better than javadoc.
```
Thank you for commenting on this. I will change the API as you mentioned.